daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala

```diff
@@ -633,27 +712,35 @@ object Ast {
       val updatedRecords = templates.map {
         case (templName, template) =>
           defsMap.get(templName) match {
-            case Some(DDataType(serializable, params, DataRecord(fields, _))) =>
-              templName -> DDataType(serializable, params, DataRecord(fields, Some(template)))
+            case Some(DDataType(serializable, params, AbstractDataRecord(fields, _))) =>
+              templName -> DDataType(
+                serializable,
+                params,
+                AbstractDataRecord(fields, Some(template)))
             case _ =>
               throw PackageError(
                 s"Data type definition not found for template ${templName.toString}")
           }
       }
 
-      new Module(name, defsMap ++ updatedRecords, featureFlags)
+      new AbstractModule[E](name, defsMap ++ updatedRecords, featureFlags)
     }
   }
 
+  type Module = AbstractModule[Expr]
+  val Module = new AbstractModuleCompanion[Expr]
+
+  type ModuleInterface = AbstractModule[Unit]
+
   case class PackageMetadata(name: PackageName, version: PackageVersion)
 
-  case class Package(
-      modules: Map[ModuleName, Module],
+  case class AbstractPackage[E](
+      modules: Map[ModuleName, AbstractModule[E]],
```

Comment:
Making this covariant will permit most cases where you've written `_` to be written `Any` instead. But your choice on which you think is nicer.

I warn that while this stack will remain perfectly sane with parametric treatment of the former `Package`, `Module`, &c components, if you decide that you want nominal subtypes later, that will cause some trouble. I think you're making a statement with this code that you don't want to see nominal subtypes at all here (as if in Haskell), but if I'm mistaken, then you might want to tie some knots in this code instead.

Reply:
I prefer `_` as `Any`